SimpleWAWS/Authentication/GitHubAuthProvider.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System.Collections.Generic;
+using Newtonsoft.Json;
+using System.Globalization;
+using System.Linq;
+using System.Net;
+using System.Security.Principal;
+using System.Text;
+using System.Web;
+using System;
+
+namespace SimpleWAWS.Authentication
+{
+    public class GitHubAuthProvider : BaseAuthProvider
+    {
+        public override void AuthenticateRequest(HttpContextBase context)
+        {
+            base.AuthenticateRequest(context, TryAuthenticateGitHubRequest);
+        }
+
+        public override string GetLoginUrl(HttpContextBase context)
+        {
+            var builder = new StringBuilder();
+            builder.Append("https://github.com/login/oauth/authorize");
+            builder.AppendFormat("?client_id={0}", AuthSettings.GitHubClientId);
+            builder.Append("&scope=user:email");
+            builder.AppendFormat("&redirect_uri={0}", WebUtility.UrlEncode(string.Format(CultureInfo.InvariantCulture, "https://{0}/", context.Request.Headers["HOST"])));
+            builder.Append(LoginStateUrlFragment(context));
+            return builder.ToString();
+        }
+
+        public override bool HasToken(HttpContextBase context)
+        {
+            return context.Request.QueryString["code"] != null;
+        }
+
+        protected TokenResults TryAuthenticateGitHubRequest(HttpContextBase context)
+        {
+            var code = context.Request.QueryString["code"];
+            if (string.IsNullOrEmpty(code))
+            {
+                return TokenResults.DoesntExist;
+            }
+
+            var user = GetUserFromGraph(code, context);
+
+            if (user == null)
+            {
+                return TokenResults.ExistAndWrong;
+            }
+            context.User = user;
+            return TokenResults.ExistsAndCorrect;
+        }
+
+        private IPrincipal GetUserFromGraph(string code, HttpContextBase context)
+        {
+            var githubAccessTokenResponse = AuthUtilities.GetContentFromUrl(GetGraphUrl(code, context), "POST", jsonAccept:true);
+
+            var githubAccessToken = JsonConvert.DeserializeObject<GitHubAccessTokenResponse>(githubAccessTokenResponse);
+
+            if (string.IsNullOrEmpty(githubAccessToken.AccessToken))
+            {
+                return null;
+            }
+            //Now get user's emailid
+            var githubUserEmailsResponse = AuthUtilities.GetContentFromUrl(GetGitHubUserUrl(), addGitHubHeaders: true, AuthorizationHeader: GetGitHubAuthHeader(githubAccessToken.AccessToken));
+            var githubUserEmails = JsonConvert.DeserializeObject<IList<GitHubUserEmailResponse>>(githubUserEmailsResponse);
+            var primaryEmail = githubUserEmails.FirstOrDefault(em => em.Primary && em.Verified);
```

Comment:
AFAIK There should always be at least one email address associated with the GitHub account. I'll return null in case `primaryEmail` is null